### ISSUE TITLE
FormData is not being send

### ISSUE THREAD

**Issue by FernetB** [capacitor-community/http#213]
**Describe the bug**
I'm trying to do a post with a img file in a form data and the request is sending nothing

**Expected behavior**
To send the formData

**Screenshots**
![image](https://user-images.githubusercontent.com/47339424/147601198-bae03d73-5941-41b8-b5a6-a367fded6e85.png)

**Comment by maximov-dev** [comment]
Hey @FernetB. Have you considered using `Http.uploadFile`?

**Comment by FernetB** [comment]
Yes, but not only doesn't work but also ask the user for permissions to the file system.

**Comment by FernetB** [comment]
@thomasvidas I'm doing something wrong ? With axios this works perfectly, but I really need to use this plugin for site-cookie reasons with capacitor

**Comment by FernetB** [comment]
On web:
![image](https://user-images.githubusercontent.com/47339424/147762688-08c70b9f-1beb-4a27-b21e-5a33eedfda27.png)
On Native:
![image](https://user-images.githubusercontent.com/47339424/147762768-899eff52-3db5-42da-8f6e-f83bdb9fa5ca.png)

**Comment by thomasvidas** [comment]
Currently on Holiday, but I'm planning on fixing this in the `2.x` branch; which isn't released yet. `Http.request` doesn't work with multipart form data. If someone else wants to make a PR I can review/release but I don't have the bandwidth to work on this.

**Comment by FernetB** [comment]
Can I use `uploadFile` for this? I tried and fail but maybe I did something wrong, or is just made for another purposes?

**Comment by christulsi** [comment]
I am having the same issue, is it fixed as yet?

**Comment by FernetB** [comment]
Nop, I have to change to `axios` and use a clouflare worker to handle the `set-cookie` and `cors` problems

**Comment by vamfixed** [comment]
any updates?

**Comment by modvd** [comment]
Any news ? Is there any way besides FormData to send multiple files at once ?

**Comment by mightytyphoon** [comment]
Hi ! This is really a critical serious issue. It should be easy to use a form and send images with POST.

I won't let you all down, as I have a fix, but this issue should be resolved ASAP in my honest opinion.

So the fix idea is to get your image as dataURI in Ionic and send it as a string. This gives lot of problems really, and should be a fix only for a short period of time.

So, using this capacitor Http plugin, you will need first to convert the image in input to a DataURI. You will need fileReader and you will probably encounter another problem of capacitor, not loading the filereader properly.

So on Ionic app (interface side) you will need : 

the fn to get fileReader
```typescript
export function getFileReader(): FileReader {
  const fileReader = new FileReader();
  const zoneOriginalInstance = (fileReader as any)["__zone_symbol__originalInstance"];
  return zoneOriginalInstance || fileReader;
};
```

the fn to get the dataURI from fileReader
```typescript
export function readURL(file: any){
    return new Promise((res, rej) => {
        const reader = getFileReader();
        reader.onload = e => res(e.target.result);
        reader.onerror = e => rej(e);
        reader.readAsDataURL(file);
    });
  };
```
the function to submit your data
```typescript
  async submit(){
    const cover = this.coverInput.nativeElement.files.item(0); // get the file from the input
    const dataURI = await this.readURL(cover); // get the dataURI
    // await this.eventService.createEvent({cover: buffer.toString()}).catch(err => {alert(JSON.stringify(err));});
    await Http.request({
      method: 'post',
      url: 'your/api/url/to/post/file',
      data : {img: dataURI}
      headers: {'Content-Type': 'application/json'}
    });
  }
```
of course you will also need to get the coverInput reference, so add the @ViewChild in your .ts file
```typescript
  @ViewChild('imgInput' , { static: true }) imgInput: ElementRef<HTMLInputElement>;
```

on html side you will need the input and a button 
```html
<input #imgInput name="cover" type="file" accept="image/*">
<button (click)="submit()">submit image</button>
```


  
Then on api side, you will need to get the string from the body and to convert it to a base64 data that can be written as a file.

you will need a getDataURI fn :
```
export function getDataURI(base64: string): {type: string , ext: string , data: string}{
    let type = '';
    let ext = '';
    let data = '';
    for(let i = 0 ; i < 60 ; i++){
      if(base64[i] === ':') {
        for(let j = 1 ; j < 60 ; j++) {
          if(base64[i+j] !== '/') {
            type = type + base64[i+j];
          } else {
            break;
          }
        }
      }
      if(base64[i] === '/') {
        for(let j = 1 ; j < 60 ; j++) {
          if(base64[i+j] !== ';') {
            ext = ext + base64[i+j];
          } else {
            break;
          }
        }
      }
      if(base64[i] === ',') {
        data = base64.substr(i + 1);
        break;
      }
    }
    return {type , ext , data}
  }
```
then you will have to get the data and write them using Buffer (on nodejs app)
```typescript
      const infos = getDataURI(body.cover);
      writeFileSync(resolve('./file.jpg')  , Buffer.from(infos.data , 'base64')
```
  
Now with this you will probably get payload too large error if your server can't handle big files (statusCode: 413)

I am personally using loopback 4 + dokku (so nginx). If you get this error, you will need to make your payload bigger on your server and on your app.
  
### in my case, for nginx you need to add bigger body size : 
```
sudo nano /etc/nginx/nginx.conf
// change line tp your needs => client_max_body_size 2M;
// for example client_max_body_size 200M;
// then save file and reload nginx
nginx -s reload
```
### and then on loopback 4, in index.ts : 
```typescript
app.bind(RestBindings.REQUEST_BODY_PARSER_OPTIONS).to({limit: '200Mb'})
```

So yeah... this works but it's kinda horrible to just upload a file... And this will get out of hand with many files, you will have to handle a sort of lazy uploading. You could also try to stream files, or use websockets to send chunks... but yeah, this shouldn't be normal usage of Http imo.

**Comment by mightytyphoon** [comment]
> Currently on Holiday, but I'm planning on fixing this in the `2.x` branch; which isn't released yet. `Http.request` doesn't work with multipart form data. If someone else wants to make a PR I can review/release but I don't have the bandwidth to work on this.

@thomasvidas Hi Thomas ! Could we have some infos about what could be wrong in the capacitor Http plugin not sending properly FormData with files using POST method ? So we can try to fix the problem and send a pull request ? 

Thank you a lot.

**Comment by FernetB** [comment]
You can't send form data because it's not yet implemented. We need to wait
until he or someone who knows Java and Swift adds it

**Comment by mightytyphoon** [comment]
@FernetB I could give it a go. (I need to put the project I am working on in production in about 3 months, and I don't want to use my fix to send files in prod)

I bet for Android I need to do something with this file : https://github.com/capacitor-community/http/blob/main/android/src/main/java/com/getcapacitor/plugin/http/FormUploader.java
and for iOS : https://github.com/capacitor-community/http/blob/main/ios/Plugin/HttpRequestHandler.swift

What I want is just someone to explain the possible problem(s) to try to fix it, this will help me to not lose too much time searching for the logic in the code.

I am into typescript and C++ everyday, but I am pretty language agnostic, so I could be able to understand how to send a formData using Java & Swift.

Here is a possible Java implementation : https://stackoverflow.com/questions/16797468/how-to-send-a-multipart-form-data-post-in-android-with-volley
And here is a possible Swift implementation : https://stackoverflow.com/questions/43492089/how-to-send-form-data-in-post-request-in-swift-3

**Comment by mightytyphoon** [comment]
Hi, I finally ended using Axios to use cookies and post forms.

Bye.

**Comment by EmilMatei** [comment]
> Hi, I finally ended using Axios to use http only cookies and post forms.
> 
> Bye.

When I've read this the first time (the "ended using axios" part) I thought it's something like "I'm done with axios, enough is enough". I understood it wrong and axios is the solution. I'm leaving this comment for other non native English speakers that may get confused as well.

**Comment by modvd** [comment]
Will there be a fix for the FormData ? Axios is not a solution for some of us and using FormData seems like an essential part of a HTTP plugin.

**Comment by riderx** [comment]
Just got my whole app stop to function because i use, this lib as custom fetcher for Supabase, and they made the storage API use the custom fetcher last week.
i will have a check and do a PR maybe next week if i have time
Edit i didn't found a way to fix the module, and it's going to get deprecated in v4
in the meantime i found a solution for my use case in Supabase
```
export const useSupabase = () => {
  const options: SupabaseClientOptions = {
    fetch: (requestInfo, requestInit) => {
      const url = requestInfo.toString()
      if (requestInit?.method === 'POST' && url.includes('/storage/')) {
        return fetch(requestInfo, requestInit)
      }
      return Http.request({
        url,
        method: requestInit?.method,
        headers: requestInit?.headers as any || {},
        data: requestInit?.body,
      })
        .then((data) => {
          const type = data.headers['content-type']
          const res = type.includes('application/vnd.pgrst.object+json') ? data.data : JSON.stringify(data.data)
          const resp = new Response(res, {
            status: data.status,
            headers: data.headers,
          })
          return resp
        })
    },
  }
  return createClient(supabaseUrl, supabaseAnonKey, options)
}
```

**Comment by nhh** [comment]
+1

**Comment by MattSynaptic** [comment]
This is also a big problem for us as well. Has there been no progress yet?

@FernetB or @mightytyphoon Is there a Capacitor plugin that you used to implement Axios or just the normal JS lib?

**Comment by mightytyphoon** [comment]
> This is also a big problem for us as well. Has there been no progress yet?
> 
> 
> 
> @FernetB or @mightytyphoon Is there a Capacitor plugin that you used to implement Axios or just the normal JS lib?

Hi, I created a http service file that handles all http requests (patch, post, get, etc...) with parameters, for now I have just 3 apps using this, so I don't need to create a package for easy management but I would advise to do so. This way if one day you need to change http and use this plugin, in case it's patched, you can easily change only the file once and this will work on all your projects by a simple update.

And yes, axios, you simply need to install it in your project with npm. It works seemlessly, don't forget to use withCredentials=true to send cookies during authenticated requests.

**Comment by maudvd** [comment]
Axios was not a good solution for me in my Ionic app.
I ended up going back to the cordova http plugin "cordova-plugin-advanced-http" which works great with FormData.

**Comment by mightytyphoon** [comment]
> Axios was not a good solution for me in my Ionic app. I ended up going back to the cordova http plugin "cordova-plugin-advanced-http" which works great with FormData.

Could you explain why, please ?

**Comment by maudvd** [comment]
> > Axios was not a good solution for me in my Ionic app. I ended up going back to the cordova http plugin "cordova-plugin-advanced-http" which works great with FormData.
> 
> Could you explain why, please ?

Because of the CORS restrictions when developing and running the app on a device.

**Comment by mightytyphoon** [comment]
> > > Axios was not a good solution for me in my Ionic app. I ended up going back to the cordova http plugin "cordova-plugin-advanced-http" which works great with FormData.
> 
> > 
> 
> > Could you explain why, please ?
> 
> 
> 
> Because of the CORS restrictions when developing and running the app on a device.

Oh yeah right, I had this problem also, you have to manage them yourself on server side. Can be annoying indeed but it works.

**Comment by rsushant** [comment]
I am facing same issue with this plugin. I tried implementing **@ionic-native/http/ngx** or **cordova-plugin-advanced-http**, project builds successfully. But in execution post methods doesn't get called or throws any exception.

Please fix capacitor-comminuty/http asap!!

**Comment by mightytyphoon** [comment]
> I am facing same issue with this plugin. I tried implementing **@ionic-native/http/ngx** or **cordova-plugin-advanced-http**, project builds successfully. But in execution post methods doesn't get called or throws any exception.
> 
> Please fix capacitor-comminuty/http asap!!

Would love to make a PR and fix this, (to do so you just need to code the formData functionality both under swift of iOS and java for android, then create the interface for execution under js/typescript), but repo master isn't responding at all nor looking at PRs...

So imo best way is to use axios, which works great, but as pointed out by @modvd you will need to manage CORS server side.

**Comment by riderx** [comment]
@mightytyphoon @rsushant the repo admin is working on 100% compatible solution to integrate into Capacitor core as mentioned in the README : https://github.com/capacitor-community/http#maintence-mode
Be patient better is coming :)
There issue to follow also https://github.com/ionic-team/capacitor/issues/5145

**Comment by mightytyphoon** [comment]
> @mightytyphoon @rsushant the repo admin is working on 100% compatible solution to integrate into Capacitor core as mentioned in the README : https://github.com/capacitor-community/http#maintence-mode
> 
> Be patient better is coming :)
> 
> 

Yeah sure, the guy came like 4 months ago, just to put one line in The readme and ignore all messages and community's PRs...

**Comment by brilliantinsane** [comment]
Still nothing?
